### ISSUE TITLE
visual studio code debug code does not match gopath using golang

### ISSUE THREAD

**Issue by friends110110** [Microsoft/vscode-go#1203]
visual studio code using golang  debugger does not match the gopath path code ?

I construct up golang environment using visual studio code IDE **MAC OS**, then install the necessary tools:

    go get -v -u github.com/peterh/liner github.com/derekparker/delve/cmd/dlv
    go get -u -v github.com/nsf/gocode
    go get -u -v github.com/rogpeppe/godef
    go get -u -v github.com/golang/lint/golint
    go get -u -v github.com/lukehoban/go-find-references
    go get -u -v github.com/lukehoban/go-outline
    go get -u -v sourcegraph.com/sqs/goreturns
    go get -u -v golang.org/x/tools/cmd/gorename
    go get -u -v github.com/tpng/gopkgs
    go get -u -v github.com/newhook/go-symbols

I ever set my go path `/Users/friends/gopath` , before long I changed my gopath `/Users/friends/Document/share/gopath`. I changed the gopath **~/.bash_profile** ,**visual studio code setting** about 

 `go.gopath": "/Users/friends/Documents/VirtualMachine/share/gopath`

 when I debug my code, it tips that **can't find** the file in`/Users/friends/gopath/src/...../apiSGetChainsIds.go` , actually the file exist in `/Users/friends/Documents/VirtualMachine/share/gopath/src/..../apiSGetChainsIds.go` .It is obvious that the debugger find the previous `gopath` ,is it the golang tools bug? or something I wrong?

my usersetting is

    {
        "files.autoSave": "afterDelay",
         //"go.buildOnSave": "package",
         //"go.lintOnSave": "package",
         //"go.vetOnSave": "package",
         "go.buildFlags": [],
         "go.lintFlags": [],
         "go.vetFlags": [],
         "go.useCodeSnippetsOnFunctionSuggest": false,
         "go.formatOnSave": false,
         "go.formatTool": "goreturns",
         "editor.fontSize": 14,
         "go.goroot": "/usr/local/Cellar/go/1.8.3/libexec",
         "go.gopath": "/Users/friends/Documents/VirtualMachine/share/gopath"
          
        }

**Comment by ramya-rao-a** [comment]
- In the VS Code terminal, run `echo $GOPATH`. Is the output the same as your expected GOPATH?
- The debugger cannot read VS Code settings. So you need to set the right GOPATH as an environment variable outside VS Code or you can pass the same in the debug configuration in the launch.json file

See [Debugging-Go-code-using-VS-Code](https://github.com/Microsoft/vscode-go/wiki/Debugging-Go-code-using-VS-Code) as well

**Comment by suntong** [comment]
This is happening to me as well -- I'm using Linux, so the bug has affected both MAC OS and Linux. 

-  I'm using the latest version of the Go extension (0.6.66), and my VS Code is downloaded and installed just today. 
- My go version is `go1.9.1 linux/amd64`, and 
- My OS is 

```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 17.04
Release:        17.04
Codename:       zesty
```
- In the VS Code terminal, run `echo $GOPATH`, the output is the same as my expected GOPATH.

Only after adding the `GOPATH` as an env var in the env property in the launch.json file solved the problem.

**Comment by ramya-rao-a** [comment]
@suntong The debugger should have figured out the right GOPATH based on the package beind debugged. Can you try the below?

- Does the value in the `program` property in the launch.json file fall under your expected GOPATH?
- Remove the `GOPATH` from the env property in the debug config in the launch.json file
- Add `"trace": "verbose"` to your debug config in the launch.json file
- Start debugging
- You will see verbose logging in the debug console. You should see a line "Using GOPATH:" Does this GOPATH match what you expect?

**Comment by suntong** [comment]
Thanks for the reply, @ramya-rao-a, 

The value in the program property in my launch.json file is:

     "program": "${fileDirname}"

I don't know how to check it, but all my own code are _not_ under GOPATH, as I want to distinct my own code from those I get from `go get`. 

Maybe that's the reason `vscode-go` is confused? 

Hmmm..., actually, I thought about that, and has tried to move my code under GOPATH, but that won't work either. Maybe I only did a symlink, and that won't work. 

Will try the trace: verbose next and post back...

**Comment by suntong** [comment]
previous post updated.

**Comment by ramya-rao-a** [comment]
To summarize:

- The debugger is using the GOPATH `/path/to/repo/gitwork` which is where your code really is
- But you have opened VS Code for a folder under `/path/to/repo/go-arch` which you have sym linked to  `/path/to/repo/gitwork`
- If you set GOPATH in the `env` in launch.json to  `/path/to/repo/go-arch`, then everything is fine
- You have set GOPATH as a env variable (outside of VS Code) to `/path/to/repo/go-arch`

Is that right?

**Comment by suntong** [comment]
Yep, precisely!

**Comment by ramya-rao-a** [comment]
Few more things...

- What do you see when you run the command `Go: Current Gopath`?
- Have you by any chance modified the settings `go.gopath` or `go.inferGopath` ?

**Comment by suntong** [comment]
"_What do you see when you run the command Go: Current Gopath?_"

Not sure I understand what you mean. 

But I didn't touch `go.gopath` or `go.inferGopath`, just accepting what was pre-filled `launch.json` setting for me.

**Comment by ramya-rao-a** [comment]
Press `Ctrl+Shift+P`, the command pallet will open
Type `Go: Current Gopath`, select it from the options

**Comment by suntong** [comment]
That output is correct, `Current GOPATH: /path/to/repo/go-arch`

**Comment by ramya-rao-a** [comment]
@suntong 
ok, https://github.com/Microsoft/vscode-go/blob/0.6.66/src/debugAdapter/goDebug.ts#L235 is the reason  for the debugger trying to infer the GOPATH.

I'll look at the history and try to recall why I do that even when GOPATH is set in the env.

@friends110110 Can you tell us if you use symlinks as well?

**Comment by suntong** [comment]
Just to make it clear that it fails regardless of the symlinks. I.e., whether I started VS Code from under 
`/path/to/repo/gitwork` or `/path/to/repo/go-arch`, both cases failed.

**Comment by ramya-rao-a** [comment]
When you start VS Code from under `/path/to/repo/gitwork` what does the `Using GOPATH` log show?

**Comment by suntong** [comment]
Same, 

`Current GOPATH: /path/to/repo/go-arch`

**Comment by ramya-rao-a** [comment]
No I meant the logs from the verbose logging when debugging

**Comment by suntong** [comment]
previous reply updated.

**Comment by ramya-rao-a** [comment]
@suntong I have pushed in a fix which should help your case, can you give it a try?

You will have to
- Download https://github.com/Microsoft/vscode-go/blob/master/Go-latest.vsix
- Run `code --install-extension Go-latest.vsix`
- Reload VS Code

**Comment by suntong** [comment]
OK, I'll try it out this weekend.

**Comment by suntong** [comment]
Sorry to response late, but my debugging is not working any more. I'll open another issue instead. 

But adding `"trace": "verbose"` to my debug config in the launch.json file, and judging the output of `Using GOPATH`, it is using the correct path now. However the same fix has caused my debugging not working any more...

**Comment by Boiethios** [comment]
What's up about this? The tools are installed in `/home/myself/go` but I set my `GOPATH` to `/home/myself/.go`:

In vsc terminal:

```
$ echo $GOPATH
/home/myself/.go
```

But with the command:

```
Current GOPATH: /home/development/go
```

This is a bug, because the documentation says:

> Out of the box, the extension uses the value of the environment variable GOPATH.

I am running Debian testing.

**Comment by suntong** [comment]
@Boiethios give this a try, 
https://github.com/Microsoft/vscode-go/issues/1307#issuecomment-341926159
at least it solves the GOPATH problem. 

And Ramya is waiting for more testers too. She's been very helpful. Let us know your outcome. please post to https://github.com/Microsoft/vscode-go/issues/1307#issuecomment-341926159 if you try. 

Thx

**Comment by suntong** [comment]
@ramya-rao-a I agree that this issue should not be considered as "closed".

**Comment by ramya-rao-a** [comment]
@Boiethios Can you share your settings? Have you set any of `go.gopath`, `go.inferGopath`, `go.toolsGopath` ? Please check both your user and workspace settings
